### PR TITLE
SG-8474: Adds setup logic for distributed projects

### PR DESCRIPTION
**Pull request by manneohrstrom**
- Adds logic to the setup wizard tank command to support setting up distributed config projects.
- Adds setup project API support for distributed setups via a new `install_mode=centralized|distributed` parameter (defaults to centralized for backwards comp).
- Adds parameter to control prompts when running `tank updates` in API mode.

### CODE REVIEW

File: python/tank/commands/setup_project_core.py [hunk 1]
```diff
@@ -34,11 +35,7 @@ def run_project_setup(log, sg, setup_params):
     log.info("")
     log.info("Starting project setup.")
 
-    # get the location of the configuration
-    config_location_curr_os = setup_params.get_configuration_location(sys.platform)
-    config_location_mac = setup_params.get_configuration_location("darwin")
-    config_location_linux = setup_params.get_configuration_location("linux2")
-    config_location_win = setup_params.get_configuration_location("win32")
+    # validate existing data
```
Comment (manneohrstrom) [inline python/tank/commands/setup_project_core.py]: now handled furhter down the code.

File: python/tank/commands/setup_project_core.py [hunk 2]
```diff
@@ -185,23 +250,18 @@ def run_project_setup(log, sg, setup_params):
     config_folder = os.path.join(config_location_curr_os, "config")
     storage_roots.write(sg, config_folder, storage_roots)
 
-    # now ensure there is a tank folder in every storage
```
Comment (manneohrstrom) [inline python/tank/commands/setup_project_core.py]: dead code.

File: python/tank/commands/setup_project_params.py [hunk 3]
```diff
@@ -79,6 +80,11 @@ def __init__(self, log, sg):
         # expert setting auto path mode
         self._auto_path = False
 
+        # distribution mode, e.g. use old xxx_path fields or
+        # new descriptor/uploaded config fields when storing
+        # stuff on the pipeline config.
+        self._distribution_mode = self.CENTRALIZED_CONFIG
```
Comment (manneohrstrom) [inline python/tank/commands/setup_project_params.py]: we default to centralized for backwards comp...

File: python/tank/commands/setup_project_params.py [hunk 4]
```diff
@@ -744,14 +852,39 @@ def pre_setup_validation(self):
...
 
-        # get the location of the configuration
-        config_path_current_os = self.get_configuration_location(sys.platform)
+        # ensure an uploaded_config field exists if we are doing a distributed cfg
+        if self._distribution_mode == self.DISTRIBUTED_CONFIG:
+            # if the fields don't exist in shotgun, the fields will be not be returned
+            field_data = self._sg.schema_field_read("PipelineConfiguration")
+            if "uploaded_config" not in field_data:
+                raise TankError(
+                    "Shotgun site is missing a PipelineConfiguration.uploaded_config "
```
Comment (manneohrstrom) [inline python/tank/commands/setup_project_params.py]: centralized distribution mode will raise an exception if executed on older shotguns.

File: python/tank/commands/update.py [hunk 5]
```diff
@@ -38,29 +38,44 @@ def __init__(self):
...
+            "type": "str"
+        }
         
+        self.parameters["app_filter"] = {
+            "description": "Name of app to check.",
+            "default": "ALL",
+            "type": "str"
+        }
         
+        self.parameters["supress_prompts"] = {
```
[the remark this answers is not in the document]
Comment (jfboismenu) [reply python/tank/commands/update.py]: 2 Ps

File: python/tank/commands/setup_project.py [hunk 6]
```diff
@@ -104,20 +113,21 @@ def __init__(self):
...
-                                                         "starts up toolkit to determine where to go look for the "
-                                                         "configuration. When setting this to true, you typically "
-                                                         "only need to specify the path to the current operating "
-                                                         "system configuration."),
-                                     "default": False,
-                                     "type": "bool" }
-        
-        
+        self.parameters["auto_path"] = {
+            "description": "Deprecated. Do not use this! --- "
```
Comment (jfboismenu) [inline python/tank/commands/setup_project.py]: Maybe we should also mention this is only used with the Shotgun Desktop. centralized configs without a project are ignored by the resolver.

File: python/tank/commands/setup_project_params.py [hunk 7]
```diff
@@ -375,6 +387,84 @@ def update_storage_root(self, config_uri, root_name, storage_data):
...
+
+            # update the roots.yml file in the config to
+            # match the storage selections we have made in the wizard
+            # e.g. link up a storage name e.g. 'primary' in the config
+            # against a local storage id in Shotgun.
+            #
+            # Note: We don't write any paths to the roots.yml -
+            # these are resolved at bootstrap time.
+            roots_data = {}
+            for storage_name in self.get_required_storages():
```
Comment (jfboismenu) [inline python/tank/commands/setup_project_params.py]: Should this logic be refactored in the StorageRoots class?

File: python/tank/commands/setup_project_wizard.py [hunk 8]
```diff
@@ -396,8 +407,11 @@ def get_default_configuration_location(self):
 
         if not data:
             # we are not based on an existing project. Instead pick the last primary config
+            # that is using storages (e.g. has got tank_name set)
             data = self._sg.find_one("PipelineConfiguration",
-                                     [["code", "is", "primary"]],
+                                     [["code", "is", "primary"],
```
Comment (jfboismenu) [inline python/tank/commands/setup_project_wizard.py]: Dumb question but... is Shotgun's code field case insensitive?

File: python/tank/commands/setup_project_params.py (hunk 4 above)
Comment (jfboismenu) [reply python/tank/commands/setup_project_params.py]: Should we mention that they can contact support to get the feature turned on if on a recent enough site?

Comment (manneohrstrom) [reply python/tank/commands/setup_project_params.py]: I think that's overkill.

File: python/tank/commands/update.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (manneohrstrom) [reply python/tank/commands/update.py]: oooops!

File: python/tank/commands/setup_project.py (hunk 6 above)
Comment (manneohrstrom) [reply python/tank/commands/setup_project.py]: i added a comment!

File: python/tank/commands/setup_project_wizard.py (hunk 8 above)
Comment (manneohrstrom) [reply python/tank/commands/setup_project_wizard.py]: yes, i believe so

File: python/tank/util/filesystem.py [hunk 9]
```diff
@@ -495,3 +496,118 @@ def auto_created_yml(path):
...
+    if not os.path.isfile(path):
+        raise ValueError(
+            "The path \"%s\" is not a valid file path." % path
+        )
+
+    # get the setting
+    system = sys.platform
+
+    if system.startswith("linux"):
+        cmd_args = ["xdg-open", os.path.dirname(path)]
```
Comment (pscadding) [inline python/tank/util/filesystem.py]: One thing I didn't add which I should have done possibly was a todo or comment around the linux open explorer and select file section. I couldn't find a consistent approach in Linux to do this, so I opted to just open the folder and not select the file. I think it's possible depending on the version of Linux and the file browser you use to open and highlight a file, but Linux doesn't' seem to be very standard.

Comment (jfboismenu) [reply python/tank/util/filesystem.py]: >  Linux doesn't' seem to be very standard.

Understatement of the year. ;)

Comment (manneohrstrom) [reply python/tank/util/filesystem.py]: lol. i'll add a comment.

**Pull request merged**